source/Components/Button.cpp

```diff
@@ -0,0 +1,328 @@
+/*
+** Project FileShare, 2022
+**
+** Author Léo Lhuile
+**
+** Started on  Sun Feb 13 22:03:32 2022 Léo Lhuile
+** Last update Mon Aug 29 20:45:51 2022 Léo Lhuile
+**
+** Button.cpp : Implementation of Button class
+*/
+
+#include "Components/Button.hpp"
+#include "Debug.hpp"
+
+namespace FileShare::GUI::Components {
+    Button::Button(const char* typeName, bool initRenderer)
+        : tgui::BitmapButton(typeName, initRenderer)
+    {
+        for (const auto &component : this->m_backgroundComponent->getComponents()) {
+            auto imageComponent = std::dynamic_pointer_cast<tgui::priv::dev::ImageComponent>(component);
+            if (imageComponent) {
+                this->imageComponent = imageComponent;
+                break;
+            }
+        }
+
+        this->setSize(Size::Normal);
+        this->updateAligment();
+        this->updateStyle();
+        this->onClick(&Button::setFocused, this, false);
+    }
+
+    Button::~Button() {}
+
+    void Button::canBeActive(bool active, bool automatic)
+    {
+        if (active == this->isActive()) {
+            return;
+        }
+
+        if (active) {
+            this->activeSignalId = automatic ? this->onClick(&Button::toggleActive, this) : 1;
+        } else {
+            if (this->activeSignalId > 1) {
+                this->onClick.disconnect(this->activeSignalId);
+            }
+
+            this->activeSignalId = 0;
+        }
+    }
+
+    void Button::setSize(Size size)
+    {
+        this->size = size;
+        if (size == Size::Small) {
+            this->setHeight(24);
+            this->setImageSize(12);
+            this->setTextSize(10);
+        } else if (size == Size::Normal) {
+            this->setHeight(36);
+            this->setImageSize(15);
+            this->setTextSize(12);
+        } else if (size == Size::Large) {
+            this->setHeight(48);
+            this->setImageSize(18);
+            this->setTextSize(15);
+        }
+    }
+
+    void Button::setSize(const tgui::Layout2d& size)
+    {
+        if (this->getImageScaling() >= 0) {
+            tgui::BitmapButton::setSize(size);
+            return;
+        }
+
+        tgui::Button::setSize(size);
+        this->recalculateGlyphSize();
+    }
+
+    void Button::updateSize()
+    {
+        if (this->getImageScaling() >= 0) {
```

Comment:
What is `getImageScaling` mean ?
I don't understand why we `return` and skip the whole function logic

Reply:
The Button class was really complicated because it uses a component system that it is the only one using (since it's in beta).
By default, the image size of a Button in TGui is relative to the Button size. But I wanted to be able to assign a fixed size to the image (among other things). So I took a lot of functions from its parent to adapt its content: `getImageScaling >= 0` allows me to know if a size relative to the Button size is applied or if it's a fixed size. If it's a relative size, then I return by calling the parent function to let it take control.